### ISSUE TITLE
[Week 5][Critical] Inverted Gain/Loss Branches in withdraw_from_yield Steal Staker Yield on Every Profitable Cycle

### ISSUE THREAD

**Issue by Stoicov** [Frankcastleauditor/Solana-Audit-Arena#225]
### Finding

**Week**: 5
**Researcher**: Stoicov (Github), @GeoGen100 (X)
**Severity**: Critical
**Category**: Inverted arithmetic / Logic error
**Affected function**: `withdraw_from_yield` (`stake_v2/src/instructions/withdraw_from_yield.rs`)

### Description

Lines 76–82 of `withdraw_from_yield.rs` contain inverted gain/loss branches. When `pnl_i128 >= 0` (the strategy returned a profit), the code executes `checked_sub_u64(pool.total_sol, gain)` — **subtracting** the gain. When `pnl_i128 < 0` (a loss), it executes `checked_add_u64(pool.total_sol, loss)` — **adding** the loss back. Both branches are exactly backwards.

`pool.total_sol` is the denominator used by `sol_to_franksol` and `franksol_to_sol` to price the SOL/frankSOL exchange rate. Reducing it on a gain makes each frankSOL redeemable for less SOL; inflating it on a loss makes frankSOL appear better-backed than the vault actually is.

### Impact

Every successful yield cycle silently reduces the SOL entitlement of all frankSOL holders by the yield amount — the protocol systematically steals staker yield and discards it. Conversely, losses inflate `total_sol`, creating phantom backing that allows late unstakers to drain more SOL than the vault contains, ultimately causing insolvency.

### Proof of Concept

Run with: `cargo test -p stake_v2 pnl_gain_decreases_total_sol_and_steals_yield`

```rust
// In: programs/stake_v2/tests/test_initialize.rs

#[test]
fn pnl_gain_decreases_total_sol_and_steals_yield() {
    use stake_v2::utils::franksol_to_sol;

    // 1000 SOL backing 1000 frankSOL (1:1 peg)
    let total_sol: u64 = 1_000_000_000_000;
    let supply:    u64 = 1_000_000_000_000;

    let principal_returned: u64 = 100_000_000_000; // 100 SOL was deployed
    let total_return:        u64 = 110_000_000_000; // strategy returned 110 SOL (10 SOL gain)

    let pnl = (total_return as i128) - (principal_returned as i128);
    assert!(pnl > 0);
    let gain = u64::try_from(pnl).unwrap();

    // Buggy path: gain branch subtracts from total_sol
    let total_sol_after_buggy = total_sol - gain; // 990 SOL instead of correct 1100

    let user_franksol: u64 = 100_000_000_000; // user holds 10 % of supply
    let sol_out_buggy   = franksol_to_sol(user_franksol, total_sol_after_buggy, supply).unwrap();
    let sol_out_correct = franksol_to_sol(user_franksol, total_sol + gain,        supply).unwrap();

    println!("user receives (buggy):   {}", sol_out_buggy);
    println!("user should receive:     {}", sol_out_correct);
    println!("SOL stolen from staker:  {}", sol_out_correct - sol_out_buggy);

    // Staker loses SOL despite protocol earning yield
    assert!(sol_out_buggy < 100_000_000_000, "staker loses principal despite yield gain");
    assert_eq!(sol_out_correct, 110_000_000_000, "correct formula pays principal + proportional yield");
}
```

### Recommended Fix

```diff
  if pnl_i128 >= 0 {
      let gain = u64::try_from(pnl_i128).map_err(|_| StakeError::MathOverflow)?;
-     pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), gain)?);
+     pool.total_sol = PodU64::from(checked_add_u64(pool.total_sol.get(), gain)?);
  } else {
      let loss = u64::try_from(-pnl_i128).map_err(|_| StakeError::MathOverflow)?;
-     pool.total_sol = PodU64::from(checked_add_u64(pool.total_sol.get(), loss)?);
+     pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), loss)?);
  }
```

**Comment by Stoicov** [comment]
I apologize for this. This was a mistake, I was testing something and these issues are duplicated. This is invalid.
